server/src/main/java/org/elasticsearch/index/codec/tsdb/TSDBSyntheticIdCodec.java

```diff
@@ -0,0 +1,164 @@
+            // Ensure _id exists and not indexed
+            fi = fieldInfos.fieldInfo(SYNTHETIC_ID);
+            if (fi == null) {
+                var message = "Field [" + SYNTHETIC_ID + "] does not exist";
+                assert false : message;
+                throw new IllegalArgumentException(message);
+            }
+            if (fi.getIndexOptions() != IndexOptions.NONE) {
+                assert false;
+                throw new IllegalArgumentException("Field [" + SYNTHETIC_ID + "] has incorrect index options");
+            }
+            return fi;
+        }
+
+        @Override
+        public void write(Directory directory, SegmentInfo segmentInfo, String segmentSuffix, FieldInfos fieldInfos, IOContext context)
+            throws IOException {
+
+            final var fieldInfo = safeIdFieldInfo(fieldInfos);
+            if (SyntheticIdField.hasSyntheticIdAttributes(fieldInfo.attributes()) == false) {
+                throw new IllegalArgumentException("Field [" + SYNTHETIC_ID + "] is not synthetic");
+            }
+            delegate.write(directory, segmentInfo, segmentSuffix, fieldInfos, context);
+        }
+
+        @Override
+        public FieldInfos read(Directory directory, SegmentInfo segmentInfo, String segmentSuffix, IOContext iocontext) throws IOException {
+            final var fieldInfos = delegate.read(directory, segmentInfo, segmentSuffix, iocontext);
+
+            final var fieldInfo = safeIdFieldInfo(fieldInfos);
+            if (SyntheticIdField.hasSyntheticIdAttributes(fieldInfo.attributes()) == false) {
+                throw new IllegalArgumentException("Field [" + SYNTHETIC_ID + "] is not synthetic");
+            }
+
+            // Change the _id field index options from IndexOptions.NONE to IndexOptions.DOCS, so that terms and postings work when
+            // applying doc values updates in Lucene.
+            final var infos = new FieldInfo[fieldInfos.size()];
+            int i = 0;
+            for (FieldInfo fi : fieldInfos) {
+                if (SYNTHETIC_ID.equals(fi.getName())) {
+                    final var attributes = new HashMap<>(fieldInfo.attributes());
+
+                    // Assert that PerFieldPostingsFormat are not written to field infos on disk
+                    assert attributes.containsKey(PerFieldPostingsFormat.PER_FIELD_SUFFIX_KEY) == false;
+                    assert attributes.containsKey(PerFieldPostingsFormat.PER_FIELD_SUFFIX_KEY) == false;
+
+                    // Inject attributes so that PerFieldPostingsFormat maps the synthetic _id field to the TSDBSyntheticIdPostingsFormat
+                    // This would normally be handled transparently by PerFieldPostingsFormat, but such attributes are only added if terms
+                    // are produced during indexing, which is not the case for the synthetic _id field.
+                    attributes.put(PerFieldPostingsFormat.PER_FIELD_FORMAT_KEY, TSDBSyntheticIdPostingsFormat.FORMAT_NAME);
```

Comment:
Nice trick. I couldn't find how `TSDBSyntheticIdPostingsFormat` was used and why `PerFieldFormatSupplier` wasn't using it. But then we have to build a posting list to disk, so this makes sense now.